spacy_experimental/span_finder/span_finder_component.py

```diff
@@ -0,0 +1,298 @@
+        doc_spans.append(spans)
+    return doc_spans
+
+
+def get_span_references(docs) -> Floats2d:
+    """Create a list of reference spans for scoring"""
+    doc_spans = []
+    for doc in docs:
+        spans = set()
+        for span_key in doc.spans:
+            for span in doc.spans[span_key]:
+                spans.add((span.start, span.end))
+        doc_spans.append(spans)
+    return doc_spans
+
+
+class SpanFinder(TrainablePipe):
+    """Pipeline that learns span boundaries"""
+
+    def __init__(
+        self,
+        nlp: Language,
+        model: Model[List[Doc], Floats2d],
+        name: str = "span_finder",
+        *,
+        threshold: float = 0.5,
+        scorer: Optional[Callable],
+        candidates_key: str = "span_finder_candidates",
+    ) -> None:
+        """Initialize the span boundary detector.
+        model (thinc.api.Model): The Thinc Model powering the pipeline component.
+        name (str): The component instance name, used to add entries to the
+            losses during training.
+        threshold (float): Minimum probability to consider a prediction
+            positive.
+        scorer (Optional[Callable]): The scoring method.
+        candidates_key (str): Name of the SpanGroup the span candidates are saved to
+        """
+        self.vocab = nlp
+        self.cfg = {
+            "threshold": threshold,
+        }
+        self.model = model
+        self.name = name
+        self.scorer = scorer
+        self.candidates_key = candidates_key
+
+    def predict(self, docs: Iterable[Doc]):
+        """Apply the pipeline's model to a batch of docs, without modifying them.
+        docs (Iterable[Doc]): The documents to predict.
+        RETURNS: The models prediction for each document.
+        """
+        scores = self.model.predict(docs)
+        return scores
+
+    def set_annotations(self, docs: Iterable[Doc], scores: Floats2d) -> None:
+        """Modify a batch of Doc objects, using pre-computed scores.
+        docs (Iterable[Doc]): The documents to modify.
+        scores: The scores to set, produced by SpanCategorizer.predict.
+        """
+        lengths = [len(doc) for doc in docs]
+
+        offset = 0
+        scores_per_doc = []
+        for length in lengths:
+            scores_per_doc.append(scores[offset : offset + length])
+            offset += length
+
+        for doc, doc_scores in zip(docs, scores_per_doc):
+            starts = []
+            ends = []
+            doc.spans[self.candidates_key] = []
+            for token, token_score in zip(doc, doc_scores):
+                if token_score[0] > self.cfg["threshold"]:
+                    starts.append(token.i)
+
+                if token_score[1] > self.cfg["threshold"]:
+                    ends.append(token.i + 1)
+            for start in starts:
+                for end in ends:
+                    if start < end:
+                        doc.spans[self.candidates_key].append(doc[start:end])
```

Comment:
Doesn't this really explode for longer docs?

Reply:
You're right. I'll try to think of a way to make this more efficient!

Reply:
I've rewritten the code a bit and it's slightly faster now, I couldn't think of a better way yet.